From e7735c854880084a6d97e60465f19daa42842eff Mon Sep 17 00:00:00 2001
From: Marios Makassikis <mmakassikis@freebox.fr>
Date: Thu, 6 May 2021 11:40:02 +0900
Subject: [PATCH] cifsd: Update out_buf_len in smb2_populate_readdir_entry()
Git-commit: e7735c854880084a6d97e60465f19daa42842eff
Patch-mainline: v5.15-rc1
References: bsn#12
Modified-by-SEL: No


When processing a SMB2 QUERY_DIRECTORY request,
smb2_populate_readdir_entry() is called first to fill the dot/dotdot
entries. This moves the d_info->wptr pointer but out_buf_len remains
unchanged. As a result, reserve_populate_dentry() may end up writing
past the end of the buffer since the bounds checking is done on
invalid values.

Signed-off-by: Marios Makassikis <mmakassikis@freebox.fr>
Signed-off-by: Namjae Jeon <namjae.jeon@samsung.com>
Signed-off-by: Steve French <stfrench@microsoft.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 fs/cifsd/smb2pdu.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/fs/cifsd/smb2pdu.c b/fs/cifsd/smb2pdu.c
index fec385318ff3..54df9a30bd23 100644
--- a/fs/cifsd/smb2pdu.c
+++ b/fs/cifsd/smb2pdu.c
@@ -3333,6 +3333,7 @@ static int smb2_populate_readdir_entry(struct ksmbd_conn *conn, int info_level,
 
 	d_info->last_entry_offset = d_info->data_count;
 	d_info->data_count += next_entry_offset;
+	d_info->out_buf_len -= next_entry_offset;
 	d_info->wptr += next_entry_offset;
 	kfree(conv_name);
 
-- 
2.33.0

